### PR TITLE
[#57] Implement Actor Model

### LINKED ISSUE #57: Actor Model
Implement actor model based on the class skeleton provided [here](https://github.com/cyfronet-fid/recommender-system/blob/b3de4dc93978ba0f320515c2468974a30791a60a/recommender/engine/rl_agent/models/actor_model.py).

### PR DESCRIPTION
**Pull request by wujuu**
Closes #57

### CODE REVIEW

File: recommender/engine/rl_agent/models/actor_model.py [hunk 1]
```diff
@@ -1,23 +1,56 @@
...
+        SPE: int,
+        history_embedder: HistoryEmbedder,
+        layer_sizes: Tuple[int] = (256, 512, 256),
+    ):
+        """
+        Args:
+            K: number of services to recommend
+            SE: service embedding dimension
+            UE: user embedding dimension
+            FE: filter embedding dimension
```
Comment (JanKapala) [inline recommender/engine/rl_agent/models/actor_model.py:30]: FE == SE, so FE is unnecessary.

File: recommender/engine/rl_agent/models/actor_model.py [hunk 2]
```diff
@@ -1,23 +1,56 @@
...
+            SPE search phrase embedding dimension
```
Comment (JanKapala) [inline recommender/engine/rl_agent/models/actor_model.py:31]: Search phrase is embedded as tensor of shape [X, SPE], where X is the number of words in the search phrase string and SPE is search phrase embedding dimension (of the word). So seachphrase tensor is two dimensional - not one dimensional.

Now we can have at least two possible solutions:

1. Build some kind of RNN into Actor - to reduce search phrase's tensor dimensions to SPE
2. Build SearchPhraseEmbedderModel (that is pytorch module) - that get tensor of shape [X, SPE] from SearchPhraseEmbedder (that is not pytorch module and just transform string to tensor) and reduce its dimensions using some RNN.

Solution 1 is wrong because the state (that contains embedded search phrase tensor of shape [X, SPE]) will be also provided to CriticModel, so it will have to have its own separate RNN for reducing dimensions. It is exactly the cause of separating out HistoryEmbedder and sharing its weights between ActorModel and CriticModel - but for history tensor. Now we should do the same thing for searchphrase tensor and separate out SearchPhraseEmbedderModel that weights should be shared between ActorModel and CriticModel - so soultion 2 is the good one.

To sum it up:
We should have:

- SearchPhraseEmbedder that is responsible for embedding search phrase string into search phrase tensor of shape [X, SPE]
- SearchPhraseEmbedderModel that is reponsible for reducing dimensions from [X, SPE] to [SPE] and its weights will be shared between ActorModel and Critic Model.

File: recommender/engine/rl_agent/models/actor_model.py [hunk 3]
```diff
@@ -1,23 +1,56 @@
...
-        # the .detach will be probably needed for proper training
 
-        # TODO: layers initialization
-        self.output_layer = torch.nn.Linear(None, K*SE)
+        self.layers = [nn.Linear(SE + UE + FE + SPE, layer_sizes[0])]
+        self.layers += [
+            nn.Linear(n_size, next_n_size)
+            for n_size, next_n_size in zip(layer_sizes, layer_sizes[1:])
+        ]
+        self.layers += [(nn.Linear(layer_sizes[-1], K * SE))]
```
Comment (JanKapala) [inline recommender/engine/rl_agent/models/actor_model.py:45]: If you want to have layers in the list inside the cutome pytorch module you should use [ModuleList](https://pytorch.org/docs/stable/generated/torch.nn.ModuleList.html) or [Sequential](https://pytorch.org/docs/stable/generated/torch.nn.Sequential.html) (and pass *layers_list to it). In other case parameters (weights) of the models/layers inside self.layers list will NOT be automatically registered for the ActorModel(torch.nn.Module). As a result actor_model.parameters() will not return all submodules parameters and the model will not learn correctly.

File: recommender/engine/rl_agent/models/actor_model.py [hunk 4]
```diff
@@ -29,9 +62,12 @@ def forward(self,
 
         """
 
-        services_history = self.history_embedder(services_history)
+        embedded_services_history = self.history_embedder(services_history)
+        x = torch.cat([embedded_services_history, user, filters, search_phrase], dim=1)
+
+        for layer in self.layers:
+            x = layer(x)
+            x = F.relu(x)
```
Comment (JanKapala) [inline recommender/engine/rl_agent/models/actor_model.py:70]: You can use [Sequential](ActorModel(torch.nn.Module)) - it will automatically process `for loop` for you - without explicitly coding it.

File: recommender/engine/rl_agent/models/history_embedder.py [hunk 5]
```diff
@@ -1,30 +1,46 @@
...
 
-        # TODO: implement the rest of the initialization (layers)
-        pass
+        Args:
+            SE: service embedding dimension
+            num_layers: number of hidden layers
+            dropout: a dropout probability
+        """
+        super().__init__()
+        self.rnn = nn.RNN(
```
Comment (JanKapala) [inline recommender/engine/rl_agent/models/history_embedder.py:24]: Same as here: https://github.com/cyfronet-fid/recommender-system/issues/69#discussion_r628055428 - 
to be sure, I'd rather mention it twice: here and there.

**Review by JanKapala — CHANGES_REQUESTED** [review]

File: recommender/engine/rl_agent/models/actor_model.py (hunk 2 above)
Comment (JanKapala) [reply recommender/engine/rl_agent/models/actor_model.py:31]: I will create separate issue for SearchPhraseEmbedderModel

Comment (wujuu) [reply recommender/engine/rl_agent/models/actor_model.py:31]: I agree however we should work on better names - if HistoryEmbedder is a module, then the SearchPhraseEmbedder should also be a Module. I suggest we rename SearchPhraseEmbedder to SearchPhraseExtractor or Preprocessor

File: recommender/engine/rl_agent/models/actor_model.py [hunk 6]
```diff
@@ -1,37 +1,81 @@
...
 import torch
+from torch import nn
+
+from recommender.engine.rl_agent.models.history_model import HistoryModel
+from recommender.engine.rl_agent.models.search_phrase_model import SearchPhraseModel
 
 
-class ActorModel(torch.nn.Module):
-    def __init__(self, K: int, SE: int, history_embedder: torch.nn.Module):
+class ActorModel(nn.Module):
```
Comment (JanKapala) [inline recommender/engine/rl_agent/models/actor_model.py:13]: Please add:
- `ACTOR_1 = "actor model 1"`
- `ACTOR_2 = "actor model 2"`
constants at the top of the actor_model.py module. They will be used as names for saving and loading appropriate actor models in the RL Agent Recommender

File: recommender/engine/rl_agent/models/actor_model.py [hunk 7]
```diff
@@ -1,37 +1,81 @@
...
+            UE: user embedding dimension
+            SPE search phrase embedding dimension
+            history_embedder: pytorch module implementing history embedding
+            layer_sizes: list containing number of neurons in each hidden layer
+        """
+
         super().__init__()
         self.K = K
         self.SE = SE
         self.history_embedder = history_embedder
```
Comment (JanKapala) [inline recommender/engine/rl_agent/models/actor_model.py:39]: History embedder pytorch model and Searchphrase Embedder pytorch model should be automatically loaded from the database if provided parameters are None. But this can be done in another issue - after refactoring.

**Review by JanKapala — CHANGES_REQUESTED** [review]

File: tests/test_search_phrase_model.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+def test_search_phrase_model_proper_shape():
+    SPE = 64  # service embedding dimension
+    X = 10  # length of a sequence
+    batch_size = 32
+
+    example_tensor = torch.rand(batch_size, X, SPE)
+    history_model = HistoryModel(SPE)
+
+    assert history_model(example_tensor).shape == torch.Size([batch_size, SPE])
```
Comment (JanKapala) [inline tests/test_search_phrase_model.py:16]: Looks like the copy of the `tests/test_history_model` file with changed function's name but without changing its body and imports.

**Review by JanKapala — CHANGES_REQUESTED** [review]

**Review by JanKapala — APPROVED** [review]
